### ISSUE TITLE
Cookies.py example: browser opens after shutdown

### ISSUE THREAD

**Issue by maphew** [flexxui/flexx#525]
https://flexx.readthedocs.io/en/stable/examples/cookies_src.html - Running this example in Win10 with python 3.71 seems a little out of step: the browser window doesn't open until after Flexx exits. 

`(oh)` is a conda environment with nothing installed except Flexx and dependencies.

```
(oh) C:\Users\mattw\code\oh
> python cookies.py
[I 16:22:43 flexx.app] Asset store collected 2 new modules.
[I 16:22:43 flexx.app] Serving apps at http://localhost:49191/
[I 16:22:43 flexx.app] Asset store collected 23 new modules.
[I 16:22:44 flexx.app] Starting Flexx event loop.
# nothing happening anywhere at this point
# wait couple minutes, then manually launch Firefox and point at above url
[W 16:24:33 flexx.app] Cleared 1 old pending sessions
[I 16:24:33 flexx.app] New session Cookies ESLJzoqOLzIFgp3gh1l075Ph
# close FF tab, now this happens:
[I 16:25:57 flexx.app] Session closed Cookies ESLJzoqOLzIFgp3gh1l075Ph
[I 16:25:57 flexx.app] Stopping Flexx event loop.
                              
(oh) C:\Users\mattw\code\oh   
>
# **NOW**  Google Chrome launches, with a tab pointed at 
# http://localhost:49191/Cookies/?session_id=fzVkPxmq6zETIyCThx0NU4uC
# but it errors out because Flexx has shut down 
```
Curious things:
- the observed effects aren't exactly the same each time. 1 out of the 6 tests ran while writing up this issue _did_ launch Chrome right away, while Flexx was still running
- Firefox is my default browser, not Chrome
- refreshing the page in either browser causes Flexx to shutdown

**Comment by almarklein** [comment]
> refreshing the page in either browser causes Flexx to shutdown

Right, `flx.start()` should be used instead of `flx.run()` here. Fixed that.

> Firefox is my default browser, not Chrome

That's annoying, but I can indeed reproduce it :(  Tracking this issue at https://github.com/flexxui/webruntime/issues/11

> the observed effects aren't exactly the same each time. 1 out of the 6 tests ran while writing up this issue did launch Chrome right away, while Flexx was still running

That is weird ...  is this only the case for this example, or also for others?

**Comment by maphew** [comment]
ok, can confirm using `flx.start()` instead of run stops flexx from exiting on every browser page refresh. However chrome is still launched _after_ flexx is stopped. It doesn't seem to matter if that's done by closing the shell session or with ctrl-c.

I get a traceback when stopping with `ctrl-c`, but that's probably a different issue?

```
> python cookies.py                                                            
[I 07:00:02 flexx.app] Asset store collected 2 new modules.                    
[I 07:00:03 flexx.app] Serving apps at http://localhost:49190/                 
[I 07:00:03 flexx.app] Asset store collected 23 new modules.                   
[I 07:00:04 flexx.app] Starting Flexx event loop.                              
[I 07:00:18 flexx.app] New session Cookies xhFVauNIllqKVTqH72TRzgH6            
[I 07:00:28 flexx.app] Session closed Cookies xhFVauNIllqKVTqH72TRzgH6
### Firefox page refresh         
[I 07:00:29 flexx.app] New session Cookies kH1GPquroV0Z9vht92uKCEi2
### Ctrl-C pressed            
Traceback (most recent call last):                                             
  File "cookies.py", line 26, in <module>                                      
    flx.start()                                                                
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_funcs.py", line
 35, in start                                                                  
    server.start()                                                             
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_server.py", lin
e 127, in start                                                                
    self._loop.run_forever()                                                   
  File "c:\apps\miniconda3\envs\oh\lib\asyncio\base_events.py", line 528, in ru
n_forever                                                                      
    self._run_once()                                                           
  File "c:\apps\miniconda3\envs\oh\lib\asyncio\base_events.py", line 1728, in _
run_once                                                                       
    event_list = self._selector.select(timeout)                                
  File "c:\apps\miniconda3\envs\oh\lib\selectors.py", line 323, in select      
    r, w, _ = self._select(self._readers, self._writers, [], timeout)          
  File "c:\apps\miniconda3\envs\oh\lib\selectors.py", line 314, in _select     
    r, w, x = select.select(r, w, w, timeout)                                  
TypeError: close() takes 1 positional argument but 3 were given                
### Chrome launches
```

**Comment by maphew** [comment]
I haven't reproduced the scenario where Chrome did launch right away. I think I may have had firefox pointing at the cookies server and it executed a page refresh at the same time as launching from the command line. Or something.

**Comment by almarklein** [comment]
That's odd, and annoying for you! I have not seen or heard of this issue before. Is this only the case for this example, or also for others?

**Comment by maphew** [comment]
_I ran everything that contained `.start()` in it. Here's my results._

**adding_handlers.py, icons.py** - server doesn't quit when the app is closed; `ctrl-c` doesn't do anything (I'm using a keyboard right now that doesn't have a `break` key)

**redirect.py** -  Error in browser window after using `alt-left` to return to previous page after using [redirect] button. `Ctrl-c` doesn't exit server. After ending python.exe process (with SysInternal's `pskill`) Chrome browser launches, pointing to the now defunct localhost:49190.

> Lost connection with server: Could not launch app: RuntimeError("Asked for session id 'AG9yZBJTeAjSC5909PyhhikS', but could not find it") (1003)

console:

```
> python redirect.py
[I 22:04:08 flexx.app] Asset store collected 2 new modules.
[I 22:04:09 flexx.app] Serving apps at http://localhost:49190/
[I 22:04:09 flexx.app] Asset store collected 23 new modules.
[I 22:04:09 flexx.app] Starting Flexx event loop.
[I 22:04:33 flexx.app] New session Redirect AG9yZBJTeAjSC5909PyhhikS
[I 22:04:47 flexx.app] Session closed Redirect AG9yZBJTeAjSC5909PyhhikS
[W 22:04:57 flexx.app] Cleared 1 old pending sessions
Uncaught exception GET /flexx/ws/Redirect (::1)
HTTPServerRequest(protocol='http', host='localhost:49190', method='GET', uri='/flexx/ws/Redirect', version='HTTP/1.1', remote_ip='::1')
Traceback (most recent call last):
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\tornado\websocket.py", line 546, in _run_callback
    result = callback(*args, **kwargs)
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_tornadoserver.py", line 542, in on_message
    cookies=self.cookies)
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_app.py", line 509, in connect_client
    session_id)
RuntimeError: Asked for session id 'AG9yZBJTeAjSC5909PyhhikS', but could not find it
```

**serve_multiple.py** -- appears to be a missing requirement in setup.py: _"ModuleNotFoundError: No module named 'psutil'"_. After installing that with pip it ran fine. In contrast to the others, this server did end with `ctrl-c`, but not gracefully.

```
> python serve_multiple1.py
[I 22:14:55 flexx.app] Asset store collected 2 new modules.
[I 22:14:56 flexx.app] Serving apps at http://localhost:49190/
[I 22:14:56 flexx.app] Starting Flexx event loop.
[I 22:15:50 flexx.app] Asset store collected 29 new modules.
[I 22:15:50 flexx.app] New session Monitor gEgMUw9L0BLvvspI6ULVHmYt
[I 22:15:59 flexx.app] Session closed Monitor gEgMUw9L0BLvvspI6ULVHmYt
[I 22:16:01 flexx.app] New session Demo Q7I7VXlaLg9Kv6fEpZp6K5eP
[I 22:16:32 flexx.app] Session closed Demo Q7I7VXlaLg9Kv6fEpZp6K5eP
Traceback (most recent call last):
  File "serve_multiple1.py", line 20, in <module>
    app.start()
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_funcs.py", line 35, in start
    server.start()
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_server.py", line 127, in start
    self._loop.run_forever()
  File "c:\apps\miniconda3\envs\oh\lib\asyncio\base_events.py", line 528, in run_forever
    self._run_once()
  File "c:\apps\miniconda3\envs\oh\lib\asyncio\base_events.py", line 1728, in _run_once
    event_list = self._selector.select(timeout)
  File "c:\apps\miniconda3\envs\oh\lib\selectors.py", line 323, in select
    r, w, _ = self._select(self._readers, self._writers, [], timeout)
  File "c:\apps\miniconda3\envs\oh\lib\selectors.py", line 314, in _select
    r, w, x = select.select(r, w, w, timeout)
KeyboardInterrupt
```

**serve_ssl.py** - missing `openssl` program (not surprising on Windows).

```
python serve_ssl.py
[I 22:20:45 flexx.app] Asset store collected 2 new modules.
'openssl' is not recognized as an internal or external command,
operable program or batch file.
Traceback (most recent call last):
  File "serve_ssl.py", line 34, in <module>
    flx.start()
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_funcs.py", line 33, in start
    server = current_server()
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_server.py", line 77, in current_server
    create_server()
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_server.py", line 58, in create_server
    _current_server = TornadoServer(host, port, loop, **server_kwargs)
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_tornadoserver.py", line 56, in __init__
    super().__init__(*args, **kwargs)
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_server.py", line 109, in __init__
    self._open(host, port, **kwargs)
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\flexx\app\_tornadoserver.py", line 98, in _open
    self._server = HTTPServer(self._app, **kwargs)
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\tornado\util.py", line 312, in __new__
    instance.initialize(*args, **init_kwargs)
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\tornado\httpserver.py", line 168, in initialize
    read_chunk_size=chunk_size)
  File "c:\apps\miniconda3\envs\oh\lib\site-packages\tornado\tcpserver.py", line 129, in __init__
    self.ssl_options['certfile'])
ValueError: certfile "/tmp/self-signed.crt" does not exist
```

**threaded.py** - no apparent problem, though with the server only staying up for 5 or 6 seconds I didn't get to see what the browser might have shown.

```
> python threaded.py                                          
[I 22:22:58 flexx.app] Asset store collected 2 new modules.   ```
[I 22:22:59 flexx.app] Serving apps at http://localhost:49190/
[I 22:22:59 flexx.app] Starting Flexx event loop.             
foo changed to 0                                              
foo changed to 5                                              
foo changed to 6                                              
foo changed to 7                                              
foo changed to 8                                              
```

**Comment by almarklein** [comment]
Thanks for looking into this! You uncovered some interesting issues!

I made #529 to track the bad shutdown behavior.

The issue in the redirect example is interesting, tracking that in #530.

Aside from these issues, do all examples that use `flx.launch()` launch the browser *after* the server stops? Could you please try a few different examples, and also change `browser` to `chrome-browser` and `firefox-browser`?

**Comment by maphew** [comment]
To muddy the waters, on my work machine which is also Win10 x64 all the examples I've tested so far "just work" with respect to browser launching. I'll see if I can tease out what's different between the environments.

**Comment by maphew** [comment]
On the problem computer (Surface Pro 4, Win10 Enterprise v1803 x64) *cookies.py* has the same launch-after-exit behaviour with both `browser` and `chrome-browser`. Furthermore, new Chrome sessions can't be launched from Start menu while cookies.py is running. Using _run as administrator_ **does** launch a new chrome session. When python.exe process is ended, a new chrome launches, pointed to localhost.

`firefox-browser` continues to just work.

**Prey spotted!** Moving or renaming `%appdata%\..\local\google\chrome` folder to something else makes browser and chrome-browser work.

**Comment by maphew** [comment]
I'm pretty sure this is a chrome bug. 

Using the standard "clear cache" mechanism via browser menu's didn't work (though I kept cookies, form data and so on).

Digging in to the appdata folder and watching the most recently changed files I noticed that a file called `login` was created after cookies.py server started, and 4 chrome processes showed up in Task Manager, but no browser window showed.

On killing the python process chrome would launch, and then in top right corner where Google profile avatar is shown it said "syncing paused" (I have half a dozen profiles that I use frequently). After signing in to the profile and then closing chrome the Flexx howto examples run without issue.

For reference the path looks like `C:\Users\mattw\AppData\Local\Google\Chrome\User Data\Profile 3`.
